### DISCUSSION TITLE
Syntax for caller-visible stack setup

### DISCUSSION THREAD

**Discussion by dmchurch** [dotnet/csharplang#8159] [Language Ideas] [answered]
This is an alternate (but non-conflicting) solution for the coding problem that inspired #8156. In that issue, I mention that adding `out` parameters to a method is _nearly_ cost-free in terms of performance. This proposal, on the other hand, is _completely_ cost-free in terms of performance.

The core idea is that, just as you can annotate a local variable declaration in a method with a `ref` keyword, you will be allowed to annotate with the `out` keyword instead. The semantics for a local variable declared as `out` are as follows:

- The variable must be declared at the top level of the method (not within a loop, conditional, or block). It can be declared anywhere in the function, but code style should be that all `out`-declared variables be at the top.
- The IL stores information (name and location) about all declared `out` variables for a given method, but they do not contribute to overload resolution - in other words, they aren't part of the method's signature. Since local variables aren't allowed to shadow the names of parameters, these will be guaranteed to be distinct from all parameter names (but see below for special syntax around that).
- When invoking a method with `out` locals, the callsite _may_ use named argument syntax to declare an `out` argument with the name of one of these `out` locals, with one additional restriction: the argument _must_ use the variable declaration form. In other words, it must be `out TypeName explicitlyTypedVariable` or `out var implicitlyTypedVariable`, not `out existingLocalVariable`.
- During codegen, if the invocation is not inlined, the invoked method will allocate any `out` variables as high in the stack as possible - likely just below the return address. The calling method can then use the value of the variable after the call has returned, just like with `out` parameters. Since it is the invoked method that determines the contents and layout of this caller-visible stack area, it can use these caller-visible stack addresses the same way it can use any stack variable, without an additional layer of indirection. If the caller wishes to use those variables even past another method call, it can either copy them into its own stack frame or simply extend its own stack frame to encompass them, after the call returns.

In this scenario, a change that adds an `out` variable to a method is source-compatible, IL-compatible, and even binary-compatible, as the method's calling convention does not change at all.

One other possible addition to this protocol would be to allow an `out`-declared local variable with no initializer to share a name with a mandatory or optional parameter passed by value. This would allow for both passing a variable in _and_ capturing any mutations performed on the parameter by the invoked method. For syntax, I'd recommend using the syntax `ref TypeName varName = value` in the argument position when specifying the input value, or `out TypeName varName` in the argument position if the parameter is optional and the input value is omitted. Since this can cause the invocation to match a different method overload, an alternate syntax should also be supported, like specifying the argument twice (once by position or named to provide the input value without any `ref` or `out`, and the second by named argument using `out TypeName varName` syntax).

I would not recommend allowing this syntax on `virtual` methods, including those defined as part of an `interface`. This functionality can only be used for methods whose actual implementation can be determined at compile-time.

A modification of the `CaseInsensitiveOrderPreservingStringDictionary` from #8156 to use caller-visible stack variables rather than optional ref/out parameters would be as follows:

```csharp
public class CaseInsensitiveOrderPreservingStringDictionary<TValue>
{
    private List<string> keys = new();
    private List<TValue> values = new();

    public int Count => keys.Count;

    // A reference to a dictionary location that has just been looked up, can index this dictionary
    public readonly ref struct Slot
    {
        public required readonly string Key { get; init; }
        public required int Index { get; init; }
        public bool HasValue => Index >= 0;
    }

    public TValue this[string key]
    {
        get => TryGetSlot(key, out Slot slot) ? GetFromSlot(slot) : throw new KeyNotFoundException();
        set => this[GetSlot(key)] = value;
    }

    public TValue this[Slot slot]
    {
        get => GetFromSlot(slot);
        set => SetFromSlot(ref slot, value);
    }

    public bool ContainsKey(string key, int index = 0)
    {
        // declaring both parameters as ref-able
        out var key, index;
        GetSlot(ref key, ref index).HasValue;
    }

    public bool TryGetValue(string key, out TValue value)
    {
        // The key parameter is ref-able
        out var key;
        // The slot variable is out-able
        out Slot slot;
        GetSlot(ref key, out slot);
        if (slot.HasValue) {
            value = this[slot];
            return true;
        } else {
            value = default;
            return false;
        }
    }

    public bool TryAdd(string key, TValue value)
    {
        out var key, value;
        out Slot slot;
        GetSlot(ref key, out slot)
        if (!slot.HasValue) {
            SetFromSlot(ref slot, value);
            slot = targetSlot;
            return true;
        } else {
            value = GetFromSlot(slot);
            return false;
        }
    }

    public bool TryGetSlot(string key, out Slot slot, int index = 0)
    {
        out var key, index;
        slot = GetSlot(ref key, ref index);
        return slot.HasValue;
    }

    protected Slot GetSlot(ref string key, ref int index = 0)
    {
        key = key.ToLowerInvariant();
        index = keys.IndexOf(key, index);
        return new Slot
        {
            Key = key,
            Index = index,
        };
    }
    protected TValue GetFromSlot(Slot slot)
    {
        return slot.HasValue ? values[slot.Index] : throw new KeyNotFoundException();
    }
    protected void SetFromSlot(ref Slot slot, TValue value)
    {
        if (slot.HasValue) {
            values[slot.Index] = value;
        } else {
            slot = slot with { Index = Count };
            keys.Add(slot.Key);
            values.Add(value);
        }
    }
}
```

**Comment by HaloFour** [comment]
I really don't understand what this proposal is attempting to achieve, but the notion that somehow a local can silently leak to the outside scope and somehow not impact the signature of the method is very unappealing to me.  I don't even understand how such a thing would be achieved.  Either the `ref` is exposed from the method and thus is a part of the signature, or it's not.

Also, inlining and assumptions about the interactions between methods bodies is something the compiler very explicitly doesn't get involved with.  Every method is opaque and the IL represents the intent of the code, it doesn't shuffle stuff around for the sake of optimization because the compiler never has the full picture to do so.  Inlining and the like is a runtime concern.

**Reply by dmchurch** [reply]
> Inlining and the like is a runtime concern.

That's the case for _most_ C# code, yes. However, C# can and does get used to write code that's intended to run in performance-critical situations. This is why `unsafe` mode exists, and why the `System.Runtime.Intrinsics` namespace was added - sometimes you need to do things at the level of pointers or in terms of raw processor operations, for a very small but extremely important percentage of C# code.

> I really don't understand what this proposal is attempting to achieve,

This is a novel, low-level runtime calling convention that could only be implemented in a language like C# - one which distributes code in a semi-compiled form which then gets translated to machine code at runtime. It's understandable that it seems unfamiliar, since to my knowledge, this _hasn't_ been done before - most fully-compiled languages, like C++ or Rust, wouldn't be capable of it.

It's a little tough to explain without going into really low-level technical details, but I'll do my best. What this proposal does is it provides an _alternate return value path_ for function calls. I'm using the word "function" here rather than "method" on purpose; when I say "function call", I mean "a method invocation that is translated into an actual machine-code CALL instruction" (or whatever mnemonic is appropriate). The reason that C# works so well is because most _method invocations_ don't end up getting turned into _function calls_ at runtime. That's why there's generally no runtime performance penalty for turning a field into a property, despite the fact that the runtime is "invoking a method" to run the property getter - the method invocation gets elided by the runtime, and the backing field access gets written directly into the machine code without turning it into a _function call_.

Function calls are, in effect, the boundary where C#'s exceptional ability to optimize the runtime stops. It isn't possible to do away with function calls, and the necessities of creating a stable low-level runtime require that the data interface across function-calls be very well-defined and standardized, with the end result that there aren't that many ways to get data from one function to another.

Right now, there are only three ways to get data from a called function back to the calling function:

1. you can return a simple value, like an `int` or a `double` or a reference to any `object` or array,
2. you can return a complex value, like a `struct` or a tuple, or
3. you can assign to a `ref` parameter, an `out` parameter, or any field on a passed or returned object/array.

Meanwhile, you can pass as many simple values as you like _into_ a called function, and on modern processors it won't require writing to memory at all - the plenitude of registers available on the x86-64 architecture, for example, means _most_ arguments can simply be passed in registers.

If all you need to return from a function is a single primitive value, it will under _most_ architectures be returned directly as a register. But what happens when you need to return more than just a single primitive value from a function, and what are the performance implications?

1. You can return an `object` with the required data in its fields. If you have to instantiate a new object, this is extremely expensive for performance-critical functions; even if you already have the object available, though, it will require copying the data from the function's local stack into the heap.
2. When functions return complex values, the caller is typically required to allocate extra space in its stack frame to store the returned data. Because the returned value has a defined Type, it must conform to a specific memory layout, or the runtime would break - so that means an extra step is involved in copying the data from the called function's stack into the caller's.
3. `ref` and `out` parameters, are, in fact, _parameters_. They represent data that is passed _into_ the called function; in this case, the address of a memory location that the caller has prepared or located. Because they are passed _in_, the caller has complete freedom in where to put the memory, but in trade, (a) all the code in the called function has to use indirection to access the data, and (b) the data must be stored in actual memory, not in a CPU register.

The ideal scenario would be that you could have a "return signature" with as many "return parameters" as your call signature, and then the called function could use the same registers the caller used to send the call arguments to return values, without ever having to touch the stack.

Unfortunately, C# doesn't have a way to represent this. The closest it comes is in using tuples for return values, but because the return value has to be an actual specified type, that means that it has to conform to a memory layout and thus, has to be stored in memory.

So how else can we represent this? Well, we could invent a brand-new syntax akin to `return expr;` which returns multiple values, but that feels like it would be extremely difficult to explain and differentiate from, say, returning a tuple. It'd also be a hassle and a half figuring out what the new syntax should _be_, and how it should be consumed at the callsite. Because of course, it's important that a value-returning function return _a value_, since it's part of an expression. So, I think it's best that the `return` statement be reserved for "the thing that should be the value of this function call in its expression".

Instead, I'm borrowing on the other way to get things out of a method - `out` parameters. But the parameters represent the _call_ signature - the data going _into_ the function, even if it represents the address for an `out` argument. So, instead, they look like `out` parameter declarations but they occur _inside_ the method body, to represent that the method itself has control over their allocation address, rather than the invoker.

[ETA: One of the reasons for the callsite restrictions I mention in the OP - that it has to be a brand-new declared variable - is because on some architectures, those values might be returned in registers rather than on the stack. The runtime can spill that register to the stack if needed, but if you want to save the value to some other address, you have to do the actual variable assignment yourself.]

The syntax would certainly need workshopping - I don't know what the best syntax would be to use at the callsite, but it seems reasonable to use the `out` argument notation.

I hope that explains it a little better; let me know if you need any other clarification.

**Reply by HaloFour** [reply]
Passing values to/from a function via registers is definitely runtime optimization territory.  I would suggest that the runtime could do this already for `out` parameters and possibly tuple elements without C# changes.  Even if C# were to officially support "multiple return values" outside of tuples someday, it should be through the mechanisms provided by the runtime officially, e.g. pushing more than one value onto the evaluation stack, and that should definitely affect the signature of that method properly.

**Reply by dmchurch** [reply]
> Passing values to/from a function via registers is definitely runtime optimization territory.

Of course it is, I never claimed it wasn't. And Microsoft is positioning C# as a performant alternative to C++, _which means_ there needs to be an answer to "how do you get code to be performant at runtime". C++ is extremely good at this, so C# has a long way to go to catch up - and this is my attempt at trying to get it a little closer to parity.

> I would suggest that the runtime could do this already for out parameters and possibly tuple elements without C# changes.

I would suggest that perhaps you do some research before you make the offhand suggestion that the C# development team can work miracles that break the [known tenets of computer science](https://en.wikipedia.org/wiki/Halting_problem).

**Reply by HaloFour** [reply]
How do you think runtime optimizations work today?  The C# language emits fairly normal IL without preoptimzation and the runtime looks for common patterns and optimizes for them.  That's exactly how inlining works, and why simple properties are as efficient as fields.  The compiler is emitting the full method and IL for a method, and the runtime elides it.  The C# actively avoids this space because the more assumptions are made at the language level makes it that much harder to optimize properly at the runtime level.  And sure, if you're using JIT there's a limit to how much effort the runtime can put into optimizations, but an AOT pass can be much more aggressive.

IMO, the conversation around optimizations to pass values back via registers really should start in the runtime, and then work backwards where necessary/appropriate.

**Reply by dmchurch** [reply]
Having just spent around five days poring over IL, JIT assembly, and AOT assembly, and experimenting with how the codegen changes based on what C# was written: you're attributing mystical powers to the CLR that it just _does not have_. Since you're not inclined to drop your crusade against this, I'm dropping mine. Congrats, have fun, I'll see myself out.

**Reply by alexrp** [reply]
This is a design discussion forum for a programming language. Debate, pushback, and bikeshedding is very much the norm in such spaces; people tend to be very passionate about the design and evolution of the language they write programs in every day.

One person being unconvinced by your idea doesn't automatically mean that it deserves no further discussion, or has no merit.

**Reply by HaloFour** [reply]
I'd suggest that it's still worth that conversation.  The CLR not being capable of this optimization today doesn't mean that it couldn't be tomorrow.  Or that conversation could lead them right back here, to a language feature to help enable it.

**Reply by dmchurch** [reply]
> The CLR not being capable of this optimization today doesn't mean that it couldn't be tomorrow.

You still just don't get it. Not only the CLR but _every_ compiler is incapable of this kind of optimization without some amount of high-level language support. The CLR won't perform this optimization because it _can't_. What you're saying is akin to suggesting that maybe if computers just get a little more powerful, they'll finally be able to solve the [traveling salesman problem](https://en.wikipedia.org/wiki/Travelling_salesman_problem). It's a nonsense statement.

> One person being unconvinced by your idea doesn't automatically mean that it deserves no further discussion, or has no merit.

_I_ know my idea has merit, that's why I suggested it. I've just run out of patience for answering questions you could find the answers to in a comp sci textbook. If someone else wants to pick the idea up and run with it, be my guest.

**Reply by CyrusNajmabadi** [reply]
> Not only the CLR but every compiler is incapable of this kind of optimization without some amount of high-level language support

I'm not seeing why that's the case. Can you clarify?

> I've just run out of patience for answering questions you could find the answers to in a comp sci textbook.

@dmchurch please abide by the .net code of conduct.  I'm extremely familiar with this space (having worked on compilers for over 20 years) and I still don't really get the argument your making here.  People asking for now understanding is totally reasonable.

**Reply by dmchurch** [reply]
> @dmchurch please abide by the .net code of conduct.

If you look at the thread, you'll notice my _very first response_ to @HaloFour's dismissive comment was an assumption of good-faith and an explanation, and before I said I had run out of patience, I tried to walk away. If you're going to call me on CoC violation, please do me the courtesy of not singling me out here.

> People asking for now understanding is totally reasonable.

I'd be _delighted_ to explain things further, so long as I'm not having to justify my own existence and prove my knowledge. Can I please assume that means you're interested to engage with this idea on its merits and not just reiterate the current state of the CLR to me?

**Reply by CyrusNajmabadi** [reply]
> and before I said I had run out of patience, I tried to walk away. If you're going to call me on CoC violation, please do me the courtesy of not singling me out here.

I haven't seen any other CoC issues.  No one has been unpleasant here.  Asking questions and pushing on ideas is exactly how lang design happens.

> so long as I'm not having to justify my own existence and prove my knowledge.

This is not helpful.  No one is doing that.  You are being asked reasonable questions and being engaged by people who have a lot of experience and interest in this space.  Please recognize though that there are tons of varying opinions and ideas here, and supporting your own is part for the course here.

--

For context, we have literally thousands of issues and discussions in this repo.  Most of them involve deeper discussions from teh participants on the merits and technical capabilities and challenges.  that's *normal* :)

**Reply by dmchurch** [reply]
Yeah okay if my comment about "please can we stop fighting" gets a thumbs-down emoji less than 5 minutes after I post it, this is clearly not the online community for me. Bye.

**Reply by CyrusNajmabadi** [reply]
@dmchurch people are welcome and encouraged to weigh in here.  That includes disagreement.  Like with any professional environment there are tons of opinions and perspectives, and not everyone will agree with what you say, or potentially how you say it.  The goal here is to navigate all of that to create great ideas that incorporate all of those disparate positions into something cohesive.

**Reply by dmchurch** [reply]
> No one is doing that.

Seems unnecessary to theorize when the person themself is right here. It's probably because I'm autistic, but I usually find that asking direct questions is the easiest way to get to the truth about a thing. So, @HaloFour? Were you, at any point, genuinely interested in what I have to say, or were you just looking to explain why I'm wrong? If you _are_ in fact interested then I'll apologize, but as yet I've seen nothing to indicate that you share @CyrusNajmabadi's desire for this to be a "professional environment" where ideas can thrive in discussion.

**Reply by jnm2** [reply]
It can take a while to see this, but taking things personally is always a choice.

**Reply by HaloFour** [reply]
I am interested in bridging what I see as a gap between what I think is a fairly aggressive language proposal with the platform-specific runtime optimization justifications for it.  If you eliminate the runtime conversation, what you have is a language proposal that seems quite out of place in C#, to me.  Namely, the ability to declare a local within a method body that is somehow directly observable by the caller without being a part of the method signature.  Apart from _how_ that happens from an IL point of view, there is the question as to why such a thing should exist given the various forms of "multiple return values" from methods.

As for the runtime optimization side of it, specifically passing parameters back/forth via registers or stack addresses, I'm not suggesting that this _has_ to be something that the runtime needs to figure out based only on what the C# compiler emits today.  But there is a range of options which probably exist between the JIT/AOT trying to figure that out on it's own vs. making a new language feature that seemingly would exist specifically for that purpose.

**Reply by dmchurch** [reply]
> I am interested in bridging what I see as a gap between what I think is a fairly aggressive language proposal with the platform-specific runtime optimization justifications for it.

Okay, _that_ sounds like an expression of interest, so I'll expand.

First, though: I'm sorry for behaving in a hostile manner, and I apologize for accusing you of purposefully wasting my time. All I'm interested in is having a conversation about what might, or might not, be a useful paradigm-shift for C# and the CLR.

The fundamental concept behind what I'm calling an "out local" is that it is:

- an optional data output from a method (in addition to its return value and `ref`/`out` parameters)
- that requires no extra effort on the part of the method to produce
- and requires no extra effort on the part of the caller to consume
- which can be fully ignored by unsupporting or uninterested callers without any change in IL _or_ asm code
- and which does not conflict with the method's existing low-level calling convention, in the case that the method is emitted as an actual asm function

The actual runtime implementation of this isn't in scope for a C# language proposal, but for a language feature to even be viable, it has to at least _have_ a potential implementation. So as proof-of-concept, here are two ways a runtime _might_ choose to implement an "out local":

1. On an x64 system, a method implemented as a leaf function (one with no processor-level function calls, meaning that any C# method/property calls have been inlined away by the runtime) can freely use the RCX, RDX, R8, and R9 registers without worry that they'll be clobbered. It might arrange to leave a value of interest (that is to say, the value of an "out local") in one of those registers prior to executing the `ret` that returns control to the caller.
2. On a system with fewer free registers, or in a method with lots of register-clobbering function calls, the runtime might decide to allocate the stack variables for "out locals" at the very top of its stack frame, where it will be in a predictable stack-pointer-relative location after the function returns.

I don't know of any prior art around this concept in software, but in other disciplines, it can be thought of as the analog of:

- In electronic engineering, the practice of adding an extra output to a chip or functional block if there are free pins/traces, if it seems likely that the extra signal will be of use to other components/functional blocks
- In cooking, the practice of setting aside a portion of a prepared ingredient (a spice mix, perhaps) for use later in the recipe before using the majority of the prepared ingredient immediately
- In industrial design methodologies like [Lean Six Sigma](https://en.wikipedia.org/wiki/Lean_Six_Sigma), the practice of identifying and eliminating duplicated work within an organization

As for how well it meshes with the design of C#, I'll borrow a technique from mathematics theory and translate the problem space into an equivalent domain to examine it. In this case, I'll start with the concept of a method. A method has:

- parameters, which are a way to get data and refs _into_ a method,
- local variables, which are a way to store data _inside_ a method, and
- a return value, which is a way to get data _out_ of a method.

Take, for example, the method `List<T>.FindAll`, which I'll reproduce below:

```csharp
        public List<T> FindAll(Predicate<T> match)
        {
            if (match == null)
            {
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.match);
            }

            List<T> list = new List<T>();
            for (int i = 0; i < _size; i++)
            {
                if (match(_items[i]))
                {
                    list.Add(_items[i]);
                }
            }
            return list;
        }
```

This method has two parameters (`match`, along with the implicit `this` parameter), four local variables (`this`, `match`, `list`, and `i`), and one return value (the new `List<T>` instance). When this method is executed by the target runtime, it will perform the following steps, in order:

1. It will allocate all stack space required for its local variables
2. It will copy the input arguments into the local variables allocated for them
3. It will perform the behaviors specified by the body of the function
4. It will copy the result value (the list reference) to a place where the calling function can find it, typically the accumulator register in most ABIs

In C#, we can fairly precisely emulate this behavior using a `ref struct`:

```csharp
    public class List<T> : IList<T>, IList, IReadOnlyList<T>
    {
[...]
        public ref struct FindAll_
        {
            private List<T> @this;
            private Predicate<T> match;
            private List<T> list;
            private int i;

            public readonly List<T> Result;

            public FindAll_(List<T> @this, Predicate<T> match)
            {
                this.@this = @this;
                this.match = match;
                Result = Execute();
            }
            private List<T> Execute() {
                if (match == null) {
                    ThrowHelper.ThrowArgumentNullException(ExceptionArgument.match);
                }

                list = new List<T>();
                for (i = 0; i < @this._size; i++) {
                    if (match(@this._items[i])) {
                        list.Add(@this._items[i]);
                    }
                }
                return list;
            }
        }
[...]
    }
    public static class ListExtensions
    {
        public static List<T> FindAll<T>(this List<T> @this, Predicate<T> match) => new List<T>.FindAll_(@this, match).Result;
    }
```

When the `ListExtensions.FindAll` extension method is invoked (or when another method uses the same pattern of `new().Result`), the runtime performs the following steps, in order:

1. It will allocate all stack space required for the `List<T>.FindAll_` ref struct
2. It will copy the constructor arguments into the fields allocated for them
3. It will perform the behaviors specified by the `Execute()` method
4. It will copy the result value to the Result field, so that the calling method can access it

The new `FindAll_` has, for all intents and purposes, the same behavior as the old `FindAll`, and in fact you can see that the `FindAll_.Execute()` method is almost syntax-identical to the old `FindAll` method, but everything that was a local variable (in other words, a stack variable) in the base method is instead a struct member. What this means is that the `FindAll_` struct is, in fact, a _physical realization of the original `FindAll` method's stack frame_.

If you accept the two situations (method with local variables vs ref struct with fields) as being fundamentally congruent, then the analog of the "out locals" feature is "being allowed to alter the visibility of the other fields of the ref struct besides `Result`". In other words, the behavior is already well-defined in abstract; all that this proposal adds is a syntax that allows you to _express_ that behavior when written in method form, rather than ref struct form. (Out locals would certainly need visibility as well; the default visibility should be `private`, so they can only be consumed by other methods in the same type, but it would certainly be possible to expose one as part of a public API.)

It's the runtime's responsibility, as always, to decide exactly how to implement this during JIT or AOT, it's the C# team's responsibility to determine what the actual syntax should be, and it is the CLR team's responsibility to decide how this is represented in the IL - where the metadata is stored, and how the list of "out locals" is exposed to the rest of the runtime, given that they won't be listed in the method signature.

Now, when I say "it does not affect the method signature", I don't mean "out locals are silent and invisible and spooky action at a distance" or whatever. I'm talking about "method signature" as per ECMA-335, which, so that you don't have to go look it up, consists of:

> - a calling convention,
> - the number of generic parameters, if the method is generic,
> - [eliding a long definition that boils down to "the implicit signature of the `this` parameter"]
> - a list of zero or more parameter signatures—one for each parameter of the method—and,
> - a type signature for the result value, if one is produced.

Note that the parameter signatures mentioned here are IL parameter signatures, not C# parameter signatures. In particular, the C# keywords `in`, `out`, and `ref` all equate to a IL signature of "byref". In other words, changing a `ref` to an `out` _does not affect the method signature_ - despite the fact that the C# method declaration is, obviously, different. Out locals would in the same category: their metadata _is_ preserved in the IL, but not in a way that affects the method signature. This is important for two reasons:

1. You can define a method that implements an interface (and thus must be signature-matching), and that method can provide additional output (e.g. metadata, performance details, permissions, confidentiality tags) beyond what is specified by the interface, without breaking binary compatibility. (Note that, since virtual methods aren't supported by the current iteration of this proposal, the method would need to be marked `sealed` to allow the use of out locals.)
2. _Adding_ an "out local" to a method is never a source-breaking change - this is part of "no extra effort on the part of the method to produce". Having to add API-indirecting stub methods in order to accomodate an expanded call signature is extra effort, both on the part of the programmer and the runtime. Obviously, _removing_ an out local would be a breaking change for all of its consumers.

It's not inconceivable that "out local" support could one day be added to interfaces and virtual functions, but that's explicitly out of scope for this and I'm not going to address it. For now, the only behavior I'm defining is how an out local behaves in a concrete, non-virtual function.

**Reply by HaloFour** [reply]
I've love to hear @tannergooding 's take on this.  I understand that having a way for a method to allocate stack for a callee for helper methods to be able to allocate buffers has been something that the BCL has been interested in, maybe this fits into the same space.

I do think that this approach to passing data is something that should be encoded explicitly into the signature in some manner, as it does affect the public surface of that member.  Even if adding such a value would not be a binary breaking change for a consumer, removing one should be if the consumer is expecting it, and that is something that the runtime should enforce.

**Reply by dmchurch** [reply]
I'd be quite interested in their take as well!

And, I fully agree that the runtime should _enforce_ this behavior - all the more so because, like `stackalloc`, it's a necessary (using a definition of "necessary" that means "this optimization won't be performed otherwise") intrusion of implementation into the high-level paradigm. I'm not an IL designer, but I'd expect that the proper way to encode this would be that the defining assembly would contain a list of "optional" attributes about the method, each listing the name of an out local, its type, and its lvar index. The consuming assembly, meanwhile, would contain a list of "required" attributes about the method call that specifies the name and type of a consumed out-local. (I don't recall the exact terminology, but I do remember that the CLR has the concept of "optional vs required" metadata.) If the defining method doesn't contain those out locals with those types, the reference fails.

So, not part of the _signature_ per se, but still part of the IL-visible method definition, like "number of locals" and "max stack size". And yes, if the producer _removes_ an out local, that's an IL-breaking change, but _only_ for methods that were, in fact, using those out locals. Invocations that only use the call-and-return signature are entirely unaffected.

The one thing that the IL representation _doesn't_ have is any indication on where or how the runtime should place the value - whether it gets returned in a register, or on the stack, or something entirely different. Sort of like an optional hyphen in a word-processing document, it doesn't specify an optimization the runtime _must_ perform, it just provides a place that the runtime _could_ perform such an optimization, if it chose to do so.

**Reply by tannergooding** [reply]
Didn't read the entire thread here, but having the ability to say "return a buffer that may be allocated in the callers stack space" is special and is unlikely to ever get general purpose support. We may someday design some `_malloca` equivalent, but it would be special and wouldn't allow a caller to then safely return that buffer. For inputs, `ref/out/in T` is already memory that is outside the callee and the same is true for inputs like `Span<T>`.

Really what we end up having is out of the following possible scenarios:
1. stackalloc
2. new T[]
3. ArrayPool<T>.Rent
4. NativeMemory.Allocate

1 is simple and where most restrictions come from, you typically only want to do it if you have less than or equal to 1024 bytes needing to be allocated. 2 is also simple, it allocates but the GC will clean it up implicitly. 3 is hard because you really want to `ArrayPool<T>.Return` in the case the success path is hit, you may also want it in the case of failure in which case you need a `finally` as well. 4 is very hard because you always need to free or you have a leak.

People then up needing 1 + one of 2/3/4 and so we could expose an API like `StackallocOrRent` which returns some `TempBuffer<T>` that is disposable and implicitly convertible to span. The runtime would treat such an API as intrinsic and directly emit the relevant logic to stackalloc if less than the threshold or to fallback and allocate/rent otherwise.

**Reply by dmchurch** [reply]
> "return a buffer that may be allocated in the callers stack space"

This isn't quite it - this isn't about returning allocated memory, but about returning _values_ (which, yes, may be reference types, but _wouldn't_ include stackalloc'd buffers, which would be illegal to return even in an out local, I think). [This comment](https://github.com/dotnet/csharplang/discussions/8159#discussioncomment-9583254) has I think my best description of the problem space, but I suppose I could summarize it as "it allows methods to return values, including structs and ref structs larger than `sizeof(nint)`, without incurring a copy". The core innovation is in providing a sufficiently-restricted reference-sharing mechanism (at the high level) that a calling method can safely (and provably) interrogate and potentially adopt the stack frame of a just-returned invocation (at the low level). That's why I initially termed it "caller-visible stack setup".

**Reply by tannergooding** [reply]
Returning values is already possibly via `T M()` or `out T` if you need to return multiple. This is already the most optimal way to return values and avoids needing a custom ABI per method (which would hurt performance and the ability for the register allocator to do its job).

`out T` gives you direct access to the location and allows you to avoid the copy. But that copy is also typically non-expensive in the first place and may be more desirable than repeated access to the memory location in the case you're doing many manipulations or setting and then repeatedly reading the value.

`T` returns are already in register (depending on ABI) for sufficiently sized value types and otherwise are already handled via an implicit return buffer (basically a hidden `out T result`) parameter.

**Reply by dmchurch** [reply]
> Returning values is already possibly via T M() or out T if you need to return multiple.

Agreed, though I'll note that `T M()` is kinder to the called procedure than `out T` - with `T M()`, the method doesn't even have to spill the result to memory if it doesn't want to (and if T is small enough). With `out T`, it will always be a reference, which means the called method's asm will always be indirect and it will always be storing to memory (if this method invocation is emitted as a function call rather than an inline).

If you want to return values T, U, and V, though, then `(T, U, V) M()` isn't fantastic - again, in the case that the method becomes a procedure call. `(T, U, V)` implies a storage layout, which means - again, since returns are going by the system ABI - that in practice, you can never return more than two `int`s from a method without spilling into the stack.

By contrast, out locals explicitly _don't_ define a memory layout; two of them could be returned in registers, and one in the stack, or whatever combination. It doesn't require supporting multiple ABIs, because out locals always use whatever registers and/or memory are _left over_ by the ABI. A native-to-managed call to a method with out locals conforms to the appropriate native ABI; it's just that the native code won't be able to use the out locals. (It'd have to call to a managed trampoline, first, which could return them in a more traditional fashion.)

> handled via an implicit return buffer (basically a hidden out T result) parameter

Yes, but so far as I've observed, the runtime won't write to the return buffer until the return statement is encountered - which means that return-by-value for larger structs always involves a copy, from the method's local stack variables, and that the stack + return buffer is always twice the size it needs to be, since the lvar of the struct is in a separate memory space than the result buffer. Return-by-out-local avoids making that copy by simply guaranteeing that the result will be in a predictable location relative to the stack pointer after return from the procedure call, and the runtime can emit static offset opcodes since it knows, at the time it's compiling the calling method, what the realized out-local addresses/registers are for the called method. And if the method wanted to return a 1024-byte struct, by value, as an out local, it could simply spill it to the top of the frame, the caller could extend its stack by 1k upon return, et voilà, now the calling method owns that 1k struct. Zero runtime cost.

[ETA: if the `out local` mechanism _were_ implemented, there's no reason the runtime couldn't use the `out local` plumbing in order to opportunistically rewrite return-by-value methods to use the same zero-copy mechanism, in the case of managed-to-managed calls.]

**Reply by tannergooding** [reply, marked as answer]
> With out T, it will always be a reference, which means the called method's asm will always be indirect and it will always be storing to memory (if this method invocation is emitted as a function call rather than an inline).

Yes, which is a requirement in order to meet the platform ABI and ensure the register allocator can do its job. While you could say "this method has a custom ABI and also uses `r8` to return a second result", that is overall worse for performance in the typical case.

Spilling to the stack is not a problem, especially for almost any CPU that's shipped in the past 14 or so years. CPUs have long had support for "mirroring" stack spills into part of the internal register file set and avoid actually accessing memory unless it's truly required (because something goes and tries to actually mutate stack space it shouldn't be touching).

> Yes, but so far as I've observed, the runtime won't write to the return buffer until the return statement is encountered - which means that return-by-value for larger structs always involves a copy,

There are several cases it can, but yes in general it will copy, which isn't an issue in the vast majority of code and helps keep semantics and observability of the mutation well-formed. For some cases, the JIT handling could just be better improved and then it wouldn't be an issue.

> And if the method wanted to return a 1024-byte struct, by value, as an out local, it could simply spill it to the top of the frame, the caller could extend its stack by 1k upon return, et voilà, now the calling method owns that 1k struct. Zero runtime cost.

This doesn't work due to the caller's return address and stack pointer existing between the callee and caller's stack space.

----------------

In general what you're trying to go for here is something which is overall going to be worse for performance, not better. There's a reason that even native C/C++ compilers don't typically do these types of things.

The natural code is typically the best code, especially over time.

**Reply by dmchurch** [reply]
> In general what you're trying to go for here is something which is overall going to be worse for performance, not better.

Okay, understood! I think I've communicated the idea well enough that if you think it's a non-starter, then the smart bet is that it _is_ - you are the expert here, after all. That's good enough for me, I have no desire to go digging around in codegen to make a POC if it's not gonna be worthwhile. :smile:

Thank you so much for taking the time to look at this and think on it! I really appreciate it.

**Reply by CyrusNajmabadi** [reply]
Thanks for bringing the discussion back to a friendly and constructive place :)
